### PR TITLE
[FX-4702] Make horizontal forms label column width adjustable

### PR DESCRIPTION
**Pull request by augustobmoura**
[FX-4702]

### Description

Add customization for Form label width

### How to test

- [Temploy](https://picasso.toptal.net/fx-4702-make-horizontal-forms-label-column-width-adjustable)
- Check example on temploy

### Screenshots

No screenshots

### Development checks

- [x] Add changeset according to [guidelines](https://github.com/toptal/picasso/blob/master/docs/contribution/changeset-guidelines.md) (if needed)
- [x] Read [CONTRIBUTING.md](https://github.com/toptal/picasso/blob/master/CONTRIBUTING.md) and [Component API principles](https://github.com/toptal/picasso/blob/master/docs/contribution/component-api.md)
- [x] Make sure that additions and changes on the design follow [Toptal's BASE design](https://design.toptal.net/), and it's been already discussed with designers at #-base-core
- [x] Annotate all `props` in component with documentation
- [x] Create `examples` for component
- [x] Ensure that deployed demo has expected results and good examples
- [x] Ensure the changed/created components have not caused accessibility issues. [How to use accessibility plugin in storybook](https://github.com/toptal/picasso/blob/master/docs/contribution/accessibility.md).
- [x] Self reviewed
- [x] Covered with tests ([visual tests](https://github.com/toptal/picasso/blob/master/docs/contribution/visual-testing.md) included)

**Breaking change**

- [NA] codemod is created and showcased in the changeset
- [NA] test alpha package of Picasso in StaffPortal

> All **_development checks_** should be done and set checked to pass the
> **GitHub Bot: TODOLess** action

<details>
<summary>PR commands</summary>
<br />

List of available commands:

- `@toptal-bot run package:alpha-release` - Release alpha version
- `@toptal-anvil ping reviewers` - Ping FX team for review

</details>

<details>
<summary>PR Review Guidelines</summary>
<br />

#### When to approve? ✅

**You are OK** with merging this PR and

1. You have no extra requests.
2. You have optional requests.
   1. Add `nit:` to your comment. (ex. `nit: I'd rename this variable from makeCircle to getCircle`)

#### When to request changes? ❌

**You are not OK** with merging this PR because

1. Something is broken after the changes.
2. Acceptance criteria is not reached.
3. Code is dirty.

#### When to comment (neither ✅ nor ❌)

**You want your comments to be addressed** before merging this PR in cases like:

1. There are leftovers like unnecessary logs, comments, etc.
2. You have an opinionated comment regarding the code that requires a discussion.
3. You have questions.

#### How to handle the comments?

1. An owner of a comment is the only one who can resolve it.
2. An owner of a comment must resolve it when it's addressed.
3. A PR owner must reply with ✅ when a comment is addressed.

</details>

[FX-4702]: https://toptal-core.atlassian.net/browse/FX-4702?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by pudek357** [comment]
@toptal-bot run package:alpha-release

File: packages/picasso/src/Form/Form.tsx [hunk 1]
```diff
@@ -1,30 +1,71 @@
...
+
 export interface Props extends BaseProps, FormHTMLAttributes<HTMLFormElement> {
   /** Content of Form constructed of Form elements */
   children?: ReactNode
   /** Submit handler */
   onSubmit?: FormEventHandler<HTMLFormElement>
   /** Form layout */
   layout?: 'horizontal' | 'vertical'
+  /** Label width */
+  labelWidth?: LabelSpacing | ResponsiveLabelSpacing
```
Comment (pudek357) [inline packages/picasso/src/Form/Form.tsx:30]: When I try to set `labelWidth` as:
```
labelWidth={{ xs: 4, sm: 4, md: 3, lg: 2, xl: 1 }}
```
It does not work.
It defaults to `4`:
image

File: packages/picasso/src/Form/Form.tsx [hunk 2]
```diff
@@ -1,30 +1,71 @@
...
+  FORM_LABEL_WIDTH_CSS_VARIABLE,
+  toClassName,
+  toCssVariable,
+} from './styles'
 import { FieldsLayoutContextProvider } from '../FieldsLayout'
 
+export type LabelSpacing = 1 | 2 | 3 | 4
+
+export type ResponsiveLabelSpacing = {
+  [k in BreakpointKeys]?: LabelSpacing
```
Comment (pudek357) [inline packages/picasso/src/Form/Form.tsx:19]: Shouldn't we omit `xs` and `sm` from possible keys? As for those breakpoints we always have `vertical` layout:
https://github.com/toptal/picasso/blob/83b93ed002272368e78175e732fa8db42f60f309/packages/picasso/src/FieldsLayout/FieldsLayoutContext.tsx#L24-L29

File: packages/picasso-forms/src/Form/story/HorizontalLabelWidth.example.tsx [hunk 3]
```diff
@@ -0,0 +1,26 @@
+import { FormNonCompound as Form, Input } from '@toptal/picasso-forms'
+import React from 'react'
+
+const Example = () => {
+  return (
+    <Form
+      autoComplete='off'
+      onSubmit={values => window.alert(JSON.stringify(values, undefined, 2))}
+      layout='horizontal'
+      labelWidth={1}
```
Comment (pudek357) [inline packages/picasso-forms/src/Form/story/HorizontalLabelWidth.example.tsx:10]: Can we have an example with `labelWidth={{ md: 2, .. }}` as well?

**Comment by pudek357** [comment]
@toptal-bot run package:alpha-release

**Comment by pudek357** [comment]
@toptal-bot run package:alpha-release

File: packages/picasso/src/Form/Form.tsx (hunk 2 above)
Comment (pudek357) [reply packages/picasso/src/Form/Form.tsx]: @dmaklygin what do you think about introducing such a change?
```tsx
export type ResponsiveLabelBreapointKeys = Exclude<BreakpointKeys, 'xs' | 'sm'>

export type ResponsiveLabelSpacing = {
  [k in ResponsiveLabelBreapointKeys]?: LabelSpacing
}
```

File: packages/picasso/src/FormField/styles.ts [hunk 4]
```diff
@@ -32,14 +81,19 @@ export default () =>
 
     horizontalLayout: {
       display: 'grid',
-      gridTemplateColumns: `${horizontalLabelColumnWidth} 1fr`,
+      // --form-label-width is passed down from cascading style, in this case from Form
+      '--label-width': `calc(${horizontalLabelColumnWidth} / 4 * var(${FORM_LABEL_WIDTH_CSS_VARIABLE}, 4))`,
+      gridTemplateColumns: `var(--label-width) 1fr`,
```
Comment (pudek357) [inline packages/picasso/src/FormField/styles.ts:86]: We should also have similar dynamic variable in:
https://github.com/toptal/picasso/blob/f892b756911000ca9b35ddb32665ae22384284fb/packages/picasso/src/FormActionsContainer/styles.ts#L9

File: packages/picasso/src/FieldsLayout/FieldsLayoutContext.tsx [hunk 5]
```diff
@@ -1,33 +1,43 @@
 import React, { createContext, useContext, useMemo } from 'react'
 
 import { useBreakpoint } from '../utils'
+import type { BreakpointKeys } from '../utils'
 
-export const horizontalLabelColumnWidth = '17rem'
+export type LabelSpacing = 2 | 3 | 4
```
Comment (sashuk) [inline packages/picasso/src/FieldsLayout/FieldsLayoutContext.tsx:6]: Do you think the "spacing" could be avoided in the name (and having something like `LabelColumns` instead)? I got confused with "spacing", as it usually is about space between element, but here the actual width is meant.

Another question is about the property name – if it is taken into account only for horizontal forms, should it be named `horizontalLabelWidth`?

Comment (sashuk) [reply packages/picasso/src/FieldsLayout/FieldsLayoutContext.tsx:6]: NIT: Will it be possible to type-guard using the newly introduced prop only if `layout="horizontal"` in `<Form/>`?

File: packages/picasso/src/FieldsLayout/FieldsLayoutContext.tsx [hunk 6]
```diff
@@ -1,33 +1,43 @@
...
+}
 
 export type FieldsLayoutContextValue = {
   layout: 'horizontal' | 'vertical'
+  labelWidth: LabelSpacing | ResponsiveLabelSpacing
 }
 
 const FieldsLayoutContext = createContext<FieldsLayoutContextValue>({
   layout: 'vertical',
+  labelWidth: 3,
```
Comment (pudek357) [inline packages/picasso/src/FieldsLayout/FieldsLayoutContext.tsx:19]:
```suggestion
  labelWidth: DEFAULT_LABEL_WIDTH_VALUE,
```

File: packages/picasso/src/FieldsLayout/FieldsLayoutContext.tsx [hunk 7]
```diff
@@ -1,33 +1,43 @@
...
 
 export type FieldsLayoutContextProviderProps = {
   layout?: 'horizontal' | 'vertical'
+  labelWidth?: LabelSpacing | ResponsiveLabelSpacing
   children: React.ReactNode
 }
 
 export const FieldsLayoutContextProvider = ({
   layout = 'vertical',
+  labelWidth = 3,
```
Comment (pudek357) [inline packages/picasso/src/FieldsLayout/FieldsLayoutContext.tsx:30]:
```suggestion
  labelWidth = DEFAULT_LABEL_WIDTH_VALUE,
```

File: packages/picasso/src/FieldsLayout/FieldsLayoutContext.tsx (hunk 5 above)
Comment (dmaklygin) [reply packages/picasso/src/FieldsLayout/FieldsLayoutContext.tsx]: We have to use union type, better to not do it

**Comment by dmaklygin** [comment]
@toptal-anvil ping reviewers

File: packages/picasso-forms/src/Form/story/index.jsx [hunk 8]
```diff
@@ -168,6 +168,16 @@ types supported by picasso-forms.
     },
     'picasso-form'
   )
+  .addExample(
+    'Form/story/HorizontalLabelWidth.example.tsx',
+    {
+      title: 'Horizontal with custom label width',
+      description:
+        "Customize the label width of your horizontal form, so it doesn't has too much empty space",
```
Comment (sashuk) [inline packages/picasso-forms/src/Form/story/index.jsx:176]:
```suggestion
        "Customize the label width of your horizontal form, so it doesn't have too much empty space",
```

File: .changeset/old-students-mix.md [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+'@toptal/picasso': minor
+---
+
+---
+
+### Forms
+
+- enhanced customization in horizontal Form layout: Introducing the `labelWidth` prop for adjusting the width of label columns in horizontal layout forms.
+
+The new `labelWidth` property in the `Form` component significantly increases the flexibility of form design, allowing for more responsive and visually balanced layouts. This property can be set with either a single numeric value (e.g., `2, 3, or 4`) for a uniform width across all screen sizes, or an object mapping specific widths to breakpoints (e.g., `{ md: 4, lg: 3, xl: 2 }`).
```
Comment (sashuk) [inline .changeset/old-students-mix.md:13]: NIT
```suggestion
The new `labelWidth` property in the `Form` component significantly increases the flexibility of form design, allowing for more responsive and visually balanced layouts. This property can be set with either a single numeric value (e.g., `2`, `3`, or `4`) for a uniform width across all screen sizes, or an object mapping specific widths to breakpoints (e.g., `{ md: 4, lg: 3, xl: 2 }`).
```

**Comment by sashuk** [comment]
Should responsive values follow the mobile-first approach? After `lg: 3` is removed, I was expecting that the `md: 2` setting will be applied


https://github.com/toptal/picasso/assets/1390758/b1fdd697-67e8-4c7c-ba55-d022a7714b84

**Review by pudek357 — APPROVED** [review]
The code looks good, but I agree with @sashuk comments.
Tested on temploy 👍

**Comment by dmaklygin** [comment]
> Should responsive values follow the mobile-first approach? After `lg: 3` is removed, I was expecting that the `md: 2` setting will be applied
> 
> ```
> md -->                | lg -->                | xl -->
> Screens wider than md | Screens wider than lg | Screens wider than xl 
> ```
> 
>  Kapture.2024-02-01.at.15.47.55.mp4

@sashuk , could you please check again. I fixed it

**Review by sashuk — APPROVED** [review]
Tested on temploy, works pefrect!

**Pull request merged**